lib/index.js

```diff
@@ -241,12 +245,60 @@ Socket.prototype._read = function (n) {}
  * Socket API
  */
 
-Socket.prototype.bind = function (addr) {
-  return this.bound[addr] = this._protect(nn.Bind( this.binding, addr ));
+Socket.prototype.bind = function (addr, fn) {
+  if (fn) return asyncBindWrapper.call(this, nn.asyncBind);
+  var ret;
+  this.sync('Bind', this, addr, ret);
+  return this.bound[addr];
+  function asyncBindWrapper (asyncBind) {
+    asyncBind(this.binding, addr, cb.bind(this));
+  }
+  function cb (eid) {
+    // set `this.bound[addr]` eid before cb fires
+    this.bound[addr] = eid;
+    return fn(eid);
+  }
+}
+
+Socket.prototype.connect = function (addr, fn) {
+  if (fn) return asyncConnectWrapper.call(this, nn.asyncConnect);
+  var ret;
+  this.sync('Connect', this, addr, ret);
+  return this.connected[addr];
+  function asyncConnectWrapper (asyncConnect) {
+    asyncConnect(this.binding, addr, cb.bind(this));
+  }
+  function cb (eid) {
+    // set `this.connected[addr]` eid before cb fires
+    this.connected[addr] = eid;
+    return fn(eid);
+  }
 }
 
-Socket.prototype.connect = function (addr) {
-  return this.connected[addr] = this._protect(nn.Connect( this.binding, addr ));
+/*  deasync helpers */
+Socket.prototype.sync = deasync(block);
+function block (nnMethod, ctx, opt, ret, done) {
+  var throttle = setInterval(function(){
+    if (ret !== undefined) {
+      clearInterval(throttle);
+      done();
+    }
+  }, 10);
+  ret = ctx._protect(nn[nnMethod]( ctx.binding, opt ));
+  switch (nnMethod) {
+    case 'Bind'     : return ctx.bound[opt]       = ret;
+    case 'Connect'  : return ctx.connected[opt]   = ret;
+  }
+}
+function deasync (fn) {
+  return function () {
+    var done=false, err, res;
+    fn.apply(this, Array.prototype.slice.apply(arguments).concat(cb));
+    while (!done) nn.block();
```

Comment:
TBH I'm not sure how much spin is happening there. 

Just some common sorcery :dizzy: but definitely not deep magic :milky_way: